io/src/real_sense_2_grabber.cpp

```diff
@@ -0,0 +1,356 @@
+
+    auto vertices_ptr = points.get_vertices ();
+    auto texture_ptr = points.get_texture_coordinates ();
+
+#ifdef _OPENMP
+#pragma omp parallel for 
+#endif
+    for (int index = 0; index < cloud->points.size (); ++index)
+    {
+      auto ptr = vertices_ptr + index;
+      auto uvptr = texture_ptr + index;
+      auto& p = cloud->points[index];
+
+      p.x = ptr->x;
+      p.y = ptr->y;
+      p.z = ptr->z;
+
+      std::tie (p.r, p.g, p.b) = getTextureColor (rgb, uvptr->u, uvptr->v);
+
+      p.a = 255;
+    }
+
+    return cloud;
+  }
+
+  pcl::PointCloud<pcl::PointXYZRGBA>::Ptr 
+  RealSense2Grabber::convertRGBADepthToPointXYZRGBA (const rs2::points & points, rs2::video_frame & rgb)
+  {
+    pcl::PointCloud<pcl::PointXYZRGBA>::Ptr cloud (new pcl::PointCloud<pcl::PointXYZRGBA> ());
+
+    auto sp = points.get_profile ().as<rs2::video_stream_profile> ();
+    cloud->width = sp.width ();
+    cloud->height = sp.height ();
+    cloud->is_dense = false;
+    cloud->points.resize (points.size ());
+
+    auto vertices_ptr = points.get_vertices ();
+    auto texture_ptr = points.get_texture_coordinates ();
+
+#ifdef _OPENMP
+#pragma omp parallel for 
+#endif
+    for (int index = 0; index < cloud->points.size (); ++index)
+    {
+      auto ptr = vertices_ptr + index;
+      auto uvptr = texture_ptr + index;
+      auto& p = cloud->points[index];
+
+      p.x = ptr->x;
+      p.y = ptr->y;
+      p.z = ptr->z;
+
+      std::tie (p.r, p.g, p.b) = getTextureColor (rgb, uvptr->u, uvptr->v);
+
+      p.a = 255;
+    }
+
+    return cloud;
+  }
+
+  std::tuple<uint8_t, uint8_t, uint8_t> 
+  RealSense2Grabber::getTextureColor (rs2::video_frame &texture, float u, float v)
+  {
+    auto ptr = dynamic_cast<rs2::video_frame*>(&texture);
+
+    const int w = ptr->get_width (), h = ptr->get_height ();
+    int x = std::min (std::max (int (u*w + .5f), 0), w - 1);
+    int y = std::min (std::max (int (v*h + .5f), 0), h - 1);
+    int idx = x * ptr->get_bytes_per_pixel() + y * ptr->get_stride_in_bytes ();
+    const auto texture_data = reinterpret_cast<const uint8_t*>(ptr->get_data ());
+    return std::make_tuple (texture_data[idx], texture_data[idx + 1], texture_data[idx + 2]);
+  }
+
+  void 
+  RealSense2Grabber::threadFunction ()
+  {
+    pcl::StopWatch sw;
+
+    rs2::frameset frames;
+    rs2::depth_frame depth = NULL;
+    rs2::video_frame rgb = NULL;
+    rs2::video_frame ir = NULL;
+    rs2::points points;
+
+    while (!quit_)
+    {
+      sw.reset ();
+
+      {
+        std::lock_guard<std::mutex> guard (mutex_);
+
+        // Wait for the next set of frames from the camera
+        frames = pipe_.wait_for_frames ();
+
+        depth = frames.get_depth_frame ();
+
+        ir = frames.get_infrared_frame ();
+
+        // Generate the pointcloud and texture mappings
+        points = pc_.calculate (depth);
+
+        rgb = frames.get_color_frame ();
+
+        // Tell pointcloud object to map to this color frame
+        pc_.map_to (rgb);
```

Comment:
I think that these processing should move to each conversion functions, because there are functions that do not require those processing. (e.g. XYZ is not require rgb and ir data, and not require convert thease data to texture.)
What do you think?

```cpp
ir = frames.get_infrared_frame ();
rgb = frames.get_color_frame ();
pc_.map_to (rgb);
```

Reply:
Excellent, I'll move these to their respective convert functions.  Should help

Reply:
```       
if(signal_PointXYZI->num_slots () > 0)
          ir = frames.get_infrared_frame ();

        // Generate the pointcloud and texture mappings
        points = pc_.calculate (depth);

        if (signal_PointXYZRGB->num_slots () > 0 || signal_PointXYZRGBA->num_slots () > 0)
        {
          rgb = frames.get_color_frame ();

          // Tell pointcloud object to map to this color frame
          pc_.map_to (rgb);
        }
```